src/app/pages/storage2/modules/devices/components/devices/devices.component.ts

```diff
@@ -53,40 +54,61 @@ export class DevicesComponent implements OnInit {
       .subscribe(() => this.loadTopologyAndDisks());
   }
 
-  private createDataSource(disks: VDev[]): void {
-    this.dataSource = new IxNestedTreeDataSource(disks);
-    this.dataSource.filterPredicate = (disks, query = '') => {
-      return disks.map((disk) => {
-        return findInTree([disk], (vdev) => {
-          switch (vdev.type) {
+  private createDataSource(dataNodes: DeviceNestedDataNode[]): void {
+    this.dataSource = new IxNestedTreeDataSource(dataNodes);
+    this.dataSource.filterPredicate = (dataNodes, query = '') => {
+      return dataNodes.map((dataNode) => {
+        return findInTree([dataNode], (dataNode) => {
+          switch (dataNode.type) {
             case VDevType.Disk:
-              return vdev.disk.toLowerCase().includes(query.toLowerCase());
+              return dataNode.disk?.toLowerCase().includes(query.toLowerCase());
             case VDevType.Mirror:
-              return vdev.name.toLowerCase().includes(query.toLowerCase());
+              return dataNode.name?.toLowerCase().includes(query.toLowerCase());
           }
         });
       }).filter(Boolean);
     };
   }
 
+  private createDataNodes(topology: PoolTopology): DeviceNestedDataNode[] {
+    const dataNodes: DeviceNestedDataNode[] = [];
```

Comment:
Here are 3 reasons why I prefer not to over use dynamic approaches to coding at this method:
- `PoolTopology` has each of it's properties strictly typed, it is not declared as an array or dictionary, so it is obviously not meant to be iterated through `.map` at first place
- `Object.entries` would be maintained in strict sync with the API changes. example: what if middleware would add a new key to the object, it would potentially lead to displaying unwanted root nodes in the tree
- creating a cycle around a `topology` object will lead to a creation of an unnecessary structures/condition checks just to inject side-values such as `'Data VDEVs', 'Cache'` and will lead to unnecessary over-complications

Reply:
The second point is debatable. I believe if new group will be added we should apply the same rules as for displaying it as for existing ones. But okay explicit is better than implicit. Thanks for clarifying 👍